play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala

```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
     new SystemConfiguration(loggerFactory).configure(config.wsClientConfig.ssl)
     wsClient
   }
+
+  /**
+   * Ensures:
+   * 1. [[StandaloneWSRequest.url]] path is encoded.
+   * 2. Any query params present in the URL are moved to [[StandaloneWSRequest.queryString]].
+   */
+  @throws[IllegalArgumentException]("if the url is unrepairable")
+  private[ahc] def normalize(req: StandaloneWSRequest): StandaloneWSRequest = {
+    try {
+      // Recover from https://github.com/AsyncHttpClient/async-http-client/issues/1149
+      Uri.create(req.url)
+      if (req.uri.getQuery == null) {
+        // happy path
+        req
+      } else {
+        // valid, but move query params into the Map
+        repair(req)
+      }
+    } catch {
+      case NonFatal(_) =>
+        // URI parsing error
+        repair(req)
+    }
+  }
+
+  @throws[IllegalArgumentException]("if the url is unrepairable")
+  private def repair(req: StandaloneWSRequest): StandaloneWSRequest = {
+    try {
+      val encodedAhcUri: Uri = toUri(req)
+      val javaUri = encodedAhcUri.toJavaNetURI
+      setUri(req, encodedAhcUri.withNewQuery(null).toUrl, Option(javaUri.getQuery))
+    } catch {
+      case NonFatal(t) =>
+        throw new IllegalArgumentException(s"Invalid URL ${req.url}", t)
+    }
+  }
+
+  /**
+   * Builds an AHC [[Uri]] with all parts URL encoded.
```

Comment:
Can we specify the RFC here used for the encoding?

Reply:
I assume `UriEncoder.FIXING` returns an `org.asynchttpclient.uri.Uri` capable of producing an RFC 3986 compliant `String`. It makes no guarantees in the javadoc, but the private static constant names in `Utf8UrlEncoder` mention RFC3986. I'm not sure we can go so far as to claim that the `Uri` result conforms to the RFC though. The result seems closer to it at least.

I've updated the scaladoc of this method to say that it runs it through `UriEncoder.FIXING`, which is at least 100% accurate.